File: .agents/skills/ai-code-masker/scripts/masker.py
import sys
import os
import re
import json

KEYWORDS = {
    'java': {'abstract', 'assert', 'boolean', 'break', 'byte', 'case', 'catch', 'char', 'class', 'const', 'continue', 'default', 'do', 'double', 'else', 'enum', 'extends', 'final', 'finally', 'float', 'for', 'goto', 'if', 'implements', 'import', 'instanceof', 'int', 'interface', 'long', 'native', 'new', 'package', 'private', 'protected', 'public', 'return', 'short', 'static', 'strictfp', 'super', 'switch', 'synchronized', 'this', 'throw', 'throws', 'transient', 'try', 'void', 'volatile', 'while', 'true', 'false', 'null', 'String', 'System', 'out', 'print', 'println', 'Override', 'Deprecated', 'SuppressWarnings'},
    'python': {'False', 'None', 'True', 'and', 'as', 'assert', 'async', 'await', 'break', 'class', 'continue', 'def', 'del', 'elif', 'else', 'except', 'finally', 'for', 'from', 'global', 'if', 'import', 'in', 'is', 'lambda', 'nonlocal', 'not', 'or', 'pass', 'raise', 'return', 'try', 'while', 'with', 'yield', 'print', 'len', 'range', 'str', 'int', 'float', 'list', 'dict', 'set', 'tuple', 'bool', 'self', 'cls'},
    'js': {'break', 'case', 'catch', 'class', 'const', 'continue', 'debugger', 'default', 'delete', 'do', 'else', 'export', 'extends', 'finally', 'for', 'function', 'if', 'import', 'in', 'instanceof', 'new', 'return', 'super', 'switch', 'this', 'throw', 'try', 'typeof', 'var', 'void', 'while', 'with', 'yield', 'let', 'static', 'enum', 'await', 'async', 'true', 'false', 'null', 'undefined', 'NaN', 'Infinity', 'document', 'window', 'console', 'log', 'React', 'useState', 'useEffect', 'useCallback', 'useMemo', 'useRef', 'useContext', 'useReducer', 'require', 'module', 'exports', 'from'}
}

REGEXES = {
    'python': re.compile(
        r'(?P<string>"""[\s\S]*?"""|\'\'\'[\s\S]*?\'\'\'|"(?:\\.|[^"\\])*"|\'(?:\\.|[^\'\\])*\')'
        r'|(?P<comment>#[^\n]*)'
        r'|(?P<identifier>\b[a-zA-Z_]\w*\b)'
    ),
    'java': re.compile(
        r'(?P<string>"(?:\\.|[^"\\])*"|\'(?:\\.|[^\'\\])*\')'
        r'|(?P<comment>//[^\n]*|/\*[\s\S]*?\*/)'
        r'|(?P<identifier>\b[a-zA-Z_]\w*\b)'
    ),
    'js': re.compile(
        r'(?P<string>"(?:\\.|[^"\\])*"|\'(?:\\.|[^\'\\])*\'|`[\s\S]*?`)'
        r'|(?P<comment>//[^\n]*|/\*[\s\S]*?\*/)'
        r'|(?P<identifier>\b[a-zA-Z_]\w*\b)'
    )
}

METHOD_REGEX = re.compile(r'\s*\(')

def get_language(filepath):
    ext = os.path.splitext(filepath)[1].lower()
    if ext == '.py': return 'python'
    if ext == '.java': return 'java'
    if ext in ['.js', '.jsx', '.ts', '.tsx']: return 'js'
    return 'js' # fallback

def mask_content(content, lang='python'):
    mapping = {}
    counters = {'var': 1, 'Class': 1, 'method': 1, 'CONST': 1}
    
    def replacer(match):
        if match.group('string') is not None:
            return match.group('string')
        if match.group('comment') is not None:
            return match.group('comment')
        
        ident = match.group('identifier')
        if ident in KEYWORDS[lang]:
            return ident
            
        if ident not in mapping:
            # Determine type heuristic
            if ident.isupper() and len(ident) > 1:
                t = 'CONST'
            elif ident[0].isupper():
                t = 'Class'
            else:
                if METHOD_REGEX.match(content, match.end()):
                    t = 'method'
                else:
                    t = 'var'
                    
            masked_name = f"{t}_{counters[t]}"
            counters[t] += 1
            mapping[ident] = masked_name
            
        return mapping[ident]

    masked_content = REGEXES[lang].sub(replacer, content)
    return masked_content, mapping

def unmask_content(content, mapping, lang='python'):
    # Invert mapping: masked_name -> original_name
    inv_mapping = {v: k for k, v in mapping.items()}
    
    def replacer(match):
        if match.group('string') is not None:
            return match.group('string')
        if match.group('comment') is not None:
            return match.group('comment')
            
        ident = match.group('identifier')
        if ident in inv_mapping:
            return inv_mapping[ident]
        return ident

    unmasked_content = REGEXES[lang].sub(replacer, content)
    return unmasked_content

def mask_file(filepath):
    lang = get_language(filepath)
    with open(filepath, 'r', encoding='utf-8') as f:
        content = f.read()
    
    masked_content, mapping = mask_content(content, lang)    
    with open(masked_filepath, 'w', encoding='utf-8') as f:
        f.write(masked_content)
        
    with open(map_filepath, 'w', encoding='utf-8') as f:
        json.dump(mapping, f, indent=2)
        
    print(f"Masked file saved to: {masked_filepath}")
    print(f"Mapping saved to: {map_filepath}")

def unmask_file(filepath):
    # Expects filepath ending with .masked
    if not filepath.endswith('.masked'):
        print("Error: For unmasking, please provide the .masked file.")
        sys.exit(1)
        
    base_filepath = filepath[:-7] # remove .masked
    map_filepath = base_filepath + '.map.json'
    
    if not os.path.exists(map_filepath):
        print(f"Error: Mapping file {map_filepath} not found.")
        sys.exit(1)
        
    with open(filepath, 'r', encoding='utf-8') as f:
        masked_content = f.read()
        
    with open(map_filepath, 'r', encoding='utf-8') as f:
        mapping = json.load(f)
        
    lang = get_language(base_filepath)
    unmasked_content = unmask_content(masked_content, mapping, lang)
    
    unmasked_filepath = base_filepath + '.unmasked'
    with open(unmasked_filepath, 'w', encoding='utf-8') as f:
        f.write(unmasked_content)
        
    print(f"Unmasked file saved to: {unmasked_filepath}")

if __name__ == '__main__':
    if len(sys.argv) < 3:
        print("Usage: python masker.py <mask|unmask> <filepath>")
        sys.exit(1)
        
    action = sys.argv[1]
    filepath = sys.argv[2]
    
    if not os.path.exists(filepath):
        print(f"Error: File {filepath} not found.")
        sys.exit(1)
        
    if action == 'mask':
        mask_file(filepath)
    elif action == 'unmask':
        unmask_file(filepath)
    else:
        print("Unknown action. Use 'mask' or 'unmask'.")
        sys.exit(1)

File: api/static/js/masker.js
/**
 * AI Code Masker - Core Client-Side Logic
 * (c) 2026 Privacy-First Developer Tools
 */

const KEYWORDS = {
    java: new Set(['abstract', 'assert', 'boolean', 'break', 'byte', 'case', 'catch', 'char', 'class', 'const', 'continue', 'default', 'do', 'double', 'else', 'enum', 'extends', 'final', 'finally', 'float', 'for', 'goto', 'if', 'implements', 'import', 'instanceof', 'int', 'interface', 'long', 'native', 'new', 'package', 'private', 'protected', 'public', 'return', 'short', 'static', 'strictfp', 'super', 'switch', 'synchronized', 'this', 'throw', 'throws', 'transient', 'try', 'void', 'volatile', 'while', 'true', 'false', 'null', 'String', 'System', 'out', 'print', 'println']),
    python: new Set(['False', 'None', 'True', 'and', 'as', 'assert', 'async', 'await', 'break', 'class', 'continue', 'def', 'del', 'elif', 'else', 'except', 'finally', 'for', 'from', 'global', 'if', 'import', 'in', 'is', 'lambda', 'nonlocal', 'not', 'or', 'pass', 'raise', 'return', 'try', 'while', 'with', 'yield', 'print', 'len', 'range', 'str', 'int', 'float', 'list', 'dict', 'set', 'tuple', 'bool', 'self', 'cls']),
    js: new Set(['break', 'case', 'catch', 'class', 'const', 'continue', 'debugger', 'default', 'delete', 'do', 'else', 'export', 'extends', 'finally', 'for', 'function', 'if', 'import', 'in', 'instanceof', 'new', 'return', 'super', 'switch', 'this', 'throw', 'try', 'typeof', 'var', 'void', 'while', 'with', 'yield', 'let', 'static', 'enum', 'await', 'async', 'true', 'false', 'null', 'undefined', 'NaN', 'Infinity', 'require', 'module', 'exports', 'from'])
};

const SENSITIVE_PATTERNS = [
    { name: 'EMAIL', regex: /[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}/g },
    { name: 'AWS_KEY', regex: /AKIA[0-9A-Z]{16}/g },
    { name: 'SECRET_TOKEN', regex: /(?:token|secret|key|password|auth|api)[-_]?(?:key|token|id)?\s*[:=]\s*["']([^"']+)["']/gi },
    { name: 'IP_ADDRESS', regex: /\b\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}\b/g }
];

const TOKEN_REGEX = {
    python: /(?<string>"""[\s\S]*?"""|'''[\s\S]*?'''|"(?:\\.|[^"\\])*"|'(?:\\.|[^'\\])*')|(?<comment>#[^\n]*)|(?<identifier>\b[a-zA-Z_]\w*\b)/g,
    java: /(?<string>"(?:\\.|[^"\\])*"|'(?:\\.|[^'\\])*')|(?<comment>\/\/[^\n]*|\/\*[\s\S]*?\*\/)|(?<identifier>\b[a-zA-Z_]\w*\b)/g,
    js: /(?<string>"(?:\\.|[^"\\])*"|'(?:\\.|[^'\\])*'|`[\s\S]*?`)|(?<comment>\/\/[^\n]*|\/\*[\s\S]*?\*\/)|(?<identifier>\b[a-zA-Z_]\w*\b)/g
};

const METHOD_LOOKAHEAD = /\s*\(/;
const GENERATED_MASK = /^(?:EMAIL|AWS_KEY|SECRET_TOKEN|IP_ADDRESS|var|Class|method|CONST|CUSTOM|STRING)_\d+$/;

function escapeRegExp(value) {
    return value.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

function maskStringLiteral(literal, mapping, counters) {
    let open = literal[0];
    let close = literal[literal.length - 1];
    let body = literal.slice(1, -1);

    if ((literal.startsWith('"""') && literal.endsWith('"""')) || (literal.startsWith("'''") && literal.endsWith("'''"))) {
        open = literal.slice(0, 3);
        close = literal.slice(-3);
        body = literal.slice(3, -3);
    }

    if (!body.trim()) return literal;

    if (!mapping[body]) {
        mapping[body] = `STRING_${counters.STRING++}`;
    }

    return `${open}${mapping[body]}${close}`;
}

function restoreMappedValues(value, invMapping) {
    const entries = Object.entries(invMapping).sort((a, b) => b[0].length - a[0].length);
    let restored = value;

    for (let pass = 0; pass < 4; pass++) {
        let next = restored;
        entries.forEach(([masked, original]) => {
            next = next.replace(new RegExp(escapeRegExp(masked), 'g'), () => original);
        });

        if (next === restored) break;
        restored = next;
    }

    return restored;
}

/**
 * Main Masking Function
 */
function maskContent(content, lang = 'python', customRules = [], options = {}) {
    const settings = {
        maskIdentifiers: true,
        maskSecrets: true,
        maskStrings: false,
        ...options
    };
    const mapping = {};
    const counters = { var: 1, Class: 1, method: 1, CONST: 1, CUSTOM: 1, SECRET: 1, STRING: 1 };
    
    let processedContent = content;

    customRules.forEach(rule => {
        if (rule && !mapping[rule]) {
            mapping[rule] = `CUSTOM_${counters.CUSTOM++}`;
        }
    });
    
    if (settings.maskSecrets) {
        processedContent = SENSITIVE_PATTERNS.reduce((current, pattern) => {
            return current.replace(pattern.regex, (match, group1) => {
                const secret = group1 || match; // Group 1 if captured (for key=value), else full match
                if (!mapping[secret]) {
                    const masked = `${pattern.name}_${counters.SECRET++}`;
                    mapping[secret] = masked;
                }
                return match.replace(secret, mapping[secret]);
            });
        }, processedContent);
    }

    customRules.forEach(rule => {
        if (!rule) return;
        processedContent = processedContent.replace(new RegExp(escapeRegExp(rule), 'g'), mapping[rule]);
    });

    if (!settings.maskIdentifiers && !settings.maskStrings) {
        return { maskedCode: processedContent, mapping };
    }

    const regex = TOKEN_REGEX[lang] || TOKEN_REGEX.js;
    regex.lastIndex = 0;

    const finalResult = processedContent.replace(regex, (match, string, comment, identifier, offset) => {
        if (string !== undefined) {
            return settings.maskStrings ? maskStringLiteral(string, mapping, counters) : string;
        }
        if (comment !== undefined) {
            return comment;
        }
        if (identifier === undefined) {
            return match;
        }

        const ident = identifier;

        if (mapping[ident]) return mapping[ident];
        if (GENERATED_MASK.test(ident)) return ident;
        if (!settings.maskIdentifiers) return ident;
        if (KEYWORDS[lang]?.has(ident)) return ident;

        let type = 'var';
        if (ident.toUpperCase() === ident && ident.length > 1 && !/^\d/.test(ident)) {
            type = 'CONST';
        } else if (ident[0].toUpperCase() === ident[0]) {
            type = 'Class';
        } else {
            const nextSlice = processedContent.slice(offset + match.length);
            if (METHOD_LOOKAHEAD.test(nextSlice)) {
                type = 'method';
            }
        }

        const masked = `${type}_${counters[type]++}`;
        mapping[ident] = masked;
        return masked;
    });

    return { maskedCode: finalResult, mapping };
}

/**
 * Restoration Function
 */
function unmaskContent(content, mapping, lang = 'python') {
    const invMapping = {};
    Object.entries(mapping).forEach(([original, masked]) => {
        invMapping[masked] = original;
    });

    const regex = TOKEN_REGEX[lang] || TOKEN_REGEX.js;
    regex.lastIndex = 0;

    return content.replace(regex, (match, string, comment, identifier) => {
        if (string !== undefined || comment !== undefined) {
            return restoreMappedValues(match, invMapping);
        }
        if (identifier === undefined) {
            return match;
        }

        return invMapping[identifier] || identifier;
    });
}

// Export for use in UI or Node (if needed)
if (typeof module !== 'undefined') {
    module.exports = { maskContent, unmaskContent };
}
